scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/BucketMetadata.java

```diff
@@ -217,6 +229,38 @@ public void populateDisplayData(Builder builder) {
     builder.add(DisplayData.item("filenamePrefix", filenamePrefix));
   }
 
+  public interface BucketIdFn extends Serializable {
+    int apply(HashCode hashCode, int numBuckets);
+
+    static BucketIdFn defaultFn() {
+      return (hashCode, numBuckets) -> Math.abs(hashCode.asInt()) % numBuckets;
+    }
+
+    static BucketIdFn icebergFn() {
+      return (hashCode, numBuckets) -> (hashCode.asInt() & Integer.MAX_VALUE) % numBuckets;
+    }
+  }
+
+  public interface KeyEncoder<T> extends Serializable {
+    byte[] encode(T value);
+
+    static <T> KeyEncoder<T> defaultKeyEncoder(Coder<T> coder) {
+      return value -> {
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        try {
+          coder.encode(value, baos);
+        } catch (Exception e) {
+          throw new RuntimeException("Could not encode key " + value, e);
+        }
+        return baos.toByteArray();
```

Comment:
```suggestion
        try {
          final ByteArrayOutputStream baos = new ByteArrayOutputStream();
          coder.encode(value, baos);
          return baos.toByteArray();
        } catch (Exception e) {
          throw new RuntimeException("Could not encode key " + value, e);
        }
```
Nitpick: in general, it is better to reduce the lifespan of variables